File: implementations/python/hash_table/hash_table_mid.py
import sys

N          = 10000
TABLE_SIZE = 15001

keys     = [0] * TABLE_SIZE
vals     = [0] * TABLE_SIZE
occupied = [False] * TABLE_SIZE

def setup():
    pass  # reset handles initialization

def hash_fn(key):
    return ((key * 2654435761) & 0xFFFFFFFFFFFFFFFF) % TABLE_SIZE

def insert(key, val):
    i = hash_fn(key)
    while occupied[i]:
        if keys[i] == key:
            vals[i] = val
            return
        i = (i + 1) % TABLE_SIZE
    keys[i]     = key
    vals[i]     = val
    occupied[i] = True

def lookup(key):
    i = hash_fn(key)
    while occupied[i]:
        if keys[i] == key:
            return vals[i]
        i = (i + 1) % TABLE_SIZE
    return -1

def reset():
    for i in range(TABLE_SIZE):
        occupied[i] = False
    for i in range(N):
        insert(i * 7 + 3, i * 13 + 5)

def hash_table():
    reset()
    checksum = 0
    for i in range(N):
        checksum += lookup(i * 7 + 3)
    return checksum

def main():
    setup()

    sys.stdout.write("ready\n")
    sys.stdout.flush()

    for line in sys.stdin:
        sys.stdout.write(str(hash_table()) + "\n")
        sys.stdout.flush()

if __name__ == "__main__":
    main()

File: implementations/rust/hash_table/mid/hash_table_mid.rs
use std::io::{self, BufRead, Write};

const N:          usize = 10000;
const TABLE_SIZE: usize = 15001;

static mut KEYS:     [i64; TABLE_SIZE]  = [0; TABLE_SIZE];
static mut VALS:     [i64; TABLE_SIZE]  = [0; TABLE_SIZE];
static mut OCCUPIED: [bool; TABLE_SIZE] = [false; TABLE_SIZE];

fn setup() {
    // reset handles initialization
}

fn hash_fn(key: i64) -> usize {
    ((key as u64).wrapping_mul(2654435761) & 0xFFFFFFFFFFFFFFFF) as usize % TABLE_SIZE
}

fn insert(key: i64, val: i64) {
    unsafe {
        let mut i = hash_fn(key);
        while OCCUPIED[i] {
            if KEYS[i] == key {
                VALS[i] = val;
                return;
            }
            i = (i + 1) % TABLE_SIZE;
        }
        KEYS[i]     = key;
        VALS[i]     = val;
        OCCUPIED[i] = true;
    }
}

fn lookup(key: i64) -> i64 {
    unsafe {
        let mut i = hash_fn(key);
        while OCCUPIED[i] {
            if KEYS[i] == key {
                return VALS[i];
            }
            i = (i + 1) % TABLE_SIZE;
        }
        -1
    }
}

fn reset() {
    unsafe {
        for i in 0..TABLE_SIZE {
            OCCUPIED[i] = false;
        }
    }
    for i in 0..N {
        insert(i as i64 * 7 + 3, i as i64 * 13 + 5);
    }
}

fn hash_table() -> i64 {
    reset();
    let mut checksum: i64 = 0;
    for i in 0..N {
        checksum += lookup(i as i64 * 7 + 3);
    }
    checksum
}

fn main() {
    setup();

    println!("ready");
    io::stdout().flush().unwrap();

    let stdin = io::stdin();
    for _ in stdin.lock().lines() {
        println!("{}", hash_table());
        io::stdout().flush().unwrap();
    }
}